src/order/conditionally_complete_lattice.lean

```diff
@@ -692,3 +716,114 @@ noncomputable instance with_top.with_bot.complete_lattice {α : Type*}
+structure. -/
+noncomputable instance Ioo_has_Sup [inhabited (Ioo a b)] : has_Sup (Ioo a b) := {Sup := λ t,
+if ht : Sup (coe '' t : set α) ∈ (Ioo a b) then ⟨Sup (coe '' t : set α), ht⟩ else default (Ioo a b)}
+
+@[simp] lemma Ioo_Sup_def [inhabited (Ioo a b)] :
+  @Sup (Ioo a b) _ = λ t,
+  if ht : Sup (coe '' t : set α) ∈ (Ioo a b) then ⟨Sup (coe '' t : set α), ht⟩ else default (Ioo a b) :=
+rfl
+
+/-- The `Sup` function on a nonempty open subinterval `Ioo a b` of a conditionally complete linear
+order agrees with the `Sup` of the original order, for all nonempty bounded-above sets. -/
+lemma Ioo_Sup_within [inhabited (Ioo a b)] {s : set (Ioo a b)} (hs : s.nonempty)
+  (h_bdd : bdd_above s) :
+  Sup (coe '' s : set α) = (@Sup (Ioo a b) _ s : α) :=
+begin
+  suffices h : Sup (coe '' s) ∈ Ioo a b,
+  { simp [-mem_Ioo, dif_pos h] },
+  obtain ⟨c, hcs⟩ : ∃ c, c ∈ s := hs,
+  obtain ⟨B, hB⟩ : ∃ B, B ∈ upper_bounds s := h_bdd,
+  have ha := lt_of_lt_of_le c.2.1 ((strict_mono_coe (Ioo a b)).monotone.image_le_cSup hcs ⟨B, hB⟩),
+  have hb := lt_of_le_of_lt ((strict_mono_coe (Ioo a b)).monotone.image_cSup_le ⟨c, hcs⟩ hB) B.2.2,
+  exact ⟨ha, hb⟩,
+end
+
+end has_Sup
+
+section has_Inf
+
+/-- `has_Inf` structure on a nonempty open subinterval `Ioo a b` of a conditionally complete linear
+order. This definition is non-canonical (it uses `default (Ioo a b)`); it should be used only as
+here, as an auxiliary instance in the construction of the `conditionally_complete_linear_order`
+structure. -/
+noncomputable instance Ioo_has_Inf [inhabited (Ioo a b)] : has_Inf (Ioo a b) := {Inf := λ t,
+if ht : Inf (coe '' t : set α) ∈ (Ioo a b) then ⟨Inf (coe '' t : set α), ht⟩ else default (Ioo a b)}
+
+@[simp] lemma Ioo_Inf_def [inhabited (Ioo a b)] :
+  @Inf (Ioo a b) _ = λ t,
+  if ht : Inf (coe '' t : set α) ∈ (Ioo a b) then ⟨Inf (coe '' t : set α), ht⟩ else default (Ioo a b) :=
+rfl
+
+/-- The `Inf` function on a nonempty open subinterval `Ioo a b` of a conditionally complete linear
+order agrees with the `Inf` of the original order, for all nonempty bounded-below sets. -/
+lemma Ioo_Inf_within [inhabited (Ioo a b)] {s : set (Ioo a b)} (hs : s.nonempty)
+  (h_bdd : bdd_below s) :
+  Inf (coe '' s : set α) = (@Inf (Ioo a b) _ s : α) :=
+begin
+  suffices h : Inf (coe '' s) ∈ Ioo a b,
+  { simp [-mem_Ioo, dif_pos h] },
+  obtain ⟨c, hcs⟩ : ∃ c, c ∈ s := hs,
+  obtain ⟨B, hB⟩ : ∃ B, B ∈ lower_bounds s := h_bdd,
+  have ha := lt_of_lt_of_le B.2.1 ((strict_mono_coe (Ioo a b)).monotone.image_le_cInf ⟨c, hcs⟩ hB),
+  have hb := lt_of_le_of_lt ((strict_mono_coe (Ioo a b)).monotone.image_cInf_le hcs ⟨B, hB⟩) c.2.2,
+  exact ⟨ha, hb⟩,
+end
+
+end has_Inf
+
+/-- A nonempty open interval of a conditionally complete linear order is naturally a conditionally
+complete linear order. -/
+noncomputable instance Ioo_conditionally_complete_linear_order [inhabited (Ioo a b)] :
+  conditionally_complete_linear_order (Ioo a b) :=
```

Comment:
I think it would be nice to generalize this to an arbitrary subtype `s` for which `Sup (coe '' t : set α) ∈ s` and `Inf (coe '' t : set α) ∈ s` whenever `t` is bounded above/below within `s` (and other conditions if you need any)

Then we can instantiate that def to an instance for `Ioo`, and we could reuse it for other subsets.